### ISSUE TITLE
Kubectl top throws incomprehensible errors

### ISSUE THREAD

**Issue by vishh** [kubernetes/kubernetes#30818]
`kubectl top` throws the following error when attempting to collect metrics from a pod that was just started. 
```shell
kubectl top pods <pod-name>
error: failed to unmarshall heapster response: invalid character 'N' looking for beginning of value
```

If the pod does not exist, the error is somewhat similar. 
```shell
k top pods foo
error: failed to unmarshall heapster response: invalid character 'P' looking for beginning of value
```

cc @kubectl @pwittrock @piosz @davidopp

**Comment by piosz** [comment]
this should be fixed by bumping Heapster version - the new version has support for lists
cc @mksalawa

**Comment by asalkeld** [comment]
I think there is also a bug in DoRaw() that ignore the http StatusCode. I'll work up a PR to address that.

**Comment by mksalawa** [comment]
Thanks @asalkeld! This solved the unmarshalling error.

The error message though is still a bit confusing:
`Error from server: the server could not find the requested resource (get services http:heapster:)`
From what I understood, this is because of the proxy error handling. The flow is:
1. Heapster doesn't find the pod - returns status 404
2. The status is redirected to the API server, which makes it think that Heapster was not found, instead of the resource.

It would get fixed after turning Heapster into apiserver. I checked with @piosz, and one of the solutions would be returning a different status code, and then handling that in the metrics client, but I'm not sure if it's a good idea. What do you think?

**Comment by piosz** [comment]
Fixed in #31510
